presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/avro/AvroRowEncoder.java

```diff
@@ -0,0 +1,157 @@
+    {
+        super(session, columnHandles);
+        this.schema = requireNonNull(parsedSchema, "parsedSchema is null");
+        this.record = new GenericData.Record(parsedSchema);
+    }
+
+    @Override
+    protected void validateColumns(List<EncoderColumnHandle> columnHandles)
+    {
+        for (EncoderColumnHandle columnHandle : columnHandles) {
+            checkArgument(columnHandle.getFormatHint() == null, "unexpected format hint '%s' defined for column '%s'", columnHandle.getFormatHint(), columnHandle.getName());
+            checkArgument(columnHandle.getDataFormat() == null, "unexpected data format '%s' defined for column '%s'", columnHandle.getDataFormat(), columnHandle.getName());
+
+            checkArgument(isSupportedType(columnHandle.getType()), "unsupported column type '%s' for column '%s'", columnHandle.getType(), columnHandle.getName());
+        }
+    }
+
+    private boolean isSupportedType(Type type)
+    {
+        return isVarcharType(type) || SUPPORTED_PRIMITIVE_TYPES.contains(type);
+    }
+
+    @Override
+    protected void appendNullValue()
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), null);
+    }
+
+    @Override
+    protected void appendLong(long value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendInt(int value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendShort(short value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendByte(byte value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendDouble(double value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendFloat(float value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendBoolean(boolean value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendString(String value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendByteBuffer(ByteBuffer value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    public byte[] toByteArray()
+    {
+        // make sure entire row has been updated with new values
+        checkArgument(currentColumnIndex == columnHandles.size(), format("Missing %d columns", columnHandles.size() - currentColumnIndex + 1));
+
+        try (ByteArrayOutputStream byteArrayOuts = new ByteArrayOutputStream();
```

Comment:
This seems quite expensive for being called for each row. 
What about changing it such that `ByteArrayOutputStream bos` and `DataFileWriter dataFileWriter` are fields isn `AvroRowEncoder`.
Then in toByteArray you can just call:
```java
    bos.reset()
    dataFileWriter.append(record);
    dataFileWriter.flush();
    return bos.toByteArray();
```
 
Yet to make this change we will need to make `RowEncoder` closable, and close `ByteArrayOutputStream` and `DataFileWriter` in `AvroRowEncoder.close()`